include/nbl/video/IDescriptorPool.h

```diff
@@ -36,21 +38,21 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
             uint32_t count;
         };
 
-        struct SDescriptorOffsets
+        inline void createDescriptorSets(uint32_t count, const IGPUDescriptorSetLayout* const* layouts, core::smart_refctd_ptr<IGPUDescriptorSet>* output)
         {
-            SDescriptorOffsets()
+            for (uint32_t i = 0u; i < count; ++i)
             {
-                // The default constructor should initiailze all the offsets to an invalid value (~0u) because ~IGPUDescriptorSet relies on it to
-                // know which descriptors are present in the set and hence should be destroyed.
-                std::fill_n(data, asset::EDT_COUNT+1, ~0u);
+                auto layout = core::smart_refctd_ptr<const IGPUDescriptorSetLayout>(layouts[i]);
+                output[i++] = createDescriptorSet(std::move(layout));
             }
+        }
+        
+        core::smart_refctd_ptr<IGPUDescriptorSet> createDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout);
+        bool freeDescriptorSets(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets);
```

Comment:
having a "free" method is not compatible with reference counting.

Everything this does, needs to happen in the destructor of the `IDescriptorSet` and `IGPUDescriptorSet`.

Yes I'm aware this frees the DS one-at-a-time but its an overhead we need to accept.

Reply:
Basically `freeDescriptorSet` should be protected.

I'm aware of what you wrote

>  Descriptor sets allocated from a descriptor pool must not be freed (via vkFreeDescriptorSets) unless the pool is created with the VK_DESCRIPTOR_POOL_CREATE_FREE_DESCRIPTOR_SET_BIT bit set.

So we do the same trick as with commandbuffer, an atomic reset counter and periodic check by the descriptor set if parent pool has reset.

But as opposed to a commandbuffer, once a DS is reset its forever invalid (zombie).
_Corollary: we make the reset stamp held by the DS a `const` and we set it during construction._

We'd also make the destructor not free the descriptor set if:
- the parent pool has been reset since creation
- the parent pool doesn't have the `ECF_FREE_DESCRIPTOR_SET_BIT` (another approach: make `freeDescriptorSets` a NOOP if the flag is missing.

Reply:
> the parent pool doesn't have the ECF_FREE_DESCRIPTOR_SET_BIT (another approach: keep the NOOP behaviour of freeDescriptorSets when the flag is missing, and always call it by dtor)

I think we will have to call the dtor of `core::smart_refctd_ptr<IDescriptor>` allocated on the pool's storage no matter if the flag `ECF_FREE_DESCRIPTOR_SET_BIT` is set or not.

Resolved.

Reply:
> I think we will have to call the dtor of core::smart_refctd_ptr<IDescriptor> constructed on the pool's storage no matter if the flag ECF_FREE_DESCRIPTOR_SET_BIT is set or not.

No, because if the `vkDescriptorSet` could not have been freed, then its "using" the descriptor sets until you reset the whole pool.

We could "implicitly call reset" if we detect that there are no more descriptor sets (the very last one just killed itself) for a pool **without** the `ECF_FREE_DESCRIPTOR_SET_BIT` flag.

Reply:
Okay so you're saying that if the flag was not set then its best to assume that the pool needs the descriptor sets to be there (for whatever reason), and freeing the descriptors i.e. calling `vkFreeImage` or `vkFreeBuffer` in this case might not be valid usage.

So in conclusion, we can do this:

- When the flag is **NOT** set:
The dtors of descriptors should be called in either `IDescriptorPool::reset` or `~IDescriptorPool` (in the event that the parent pool is never reset). Following shouldn't be required then:
> We could "implicitly call reset" if we detect that there are no more descriptor sets (the very last one just killed itself) for a pool without the `ECF_FREE_DESCRIPTOR_SET_BIT` flag.

- When the flag is set:
The dtors of descriptors should be called in either `IDescriptorPool::reset` or `~IGPUDescriptorSet`.

In both cases, to prevent double frees, we will check if the pointers coming of the descriptor storage are `nullptr`.

Do you agree with this?

Reply:
> to prevent double frees, we will check if the pointers coming of the descriptor storage are nullptr.

If by this you mean, making sure the `core::smart_refctd_ptr<IDescriptor>` stored in the pool is `nullptr` after calling `~IDescriptor` on it via any mean.

Then all of your conclusions are 100% correct, except

> Following shouldn't be required then:

shouldn't be required, but you can do it anyway. Its just an extra call to `reset()` when we're basically 100% sure it is safe to do so.

Reply:
### We can throw the entire preceeding conversation in the bin 

Discord talk cleaned this up.

#### `IDescriptor` storage

We always `nullptr` the smart pointers because the Buffers, Buffer Views and Image Views can be freely destroyed as long as they wont be used in a future submit.

A destroyed `IGPUDescriptorSet` can never be used again in a future submit (as long as commandbuffer resource tracking works well to keep the DS alive), even though its `vkDescriptorSet` is still valid if it cannot be individually freed.

So there's no point keeping `IDescriptor`s of the DS alive after a destruction of the Nabla DS abstraction or reset of the pool.


#### Calling `vkFreeDescriptorSets` on the Descriptor Set is a different thing

If a pool has reset itself, its important that `~IGPUDescriptorSet` NEVER call `vkFreeDescriptorSet`.

Also its important its able to detect this to throw errors if we try to use it, since the `IDescriptorPool` can't just kill a `IGPUDescriptorSet`.

Furthermore, if pool has the FREE flag, then we neeed to call this in `~IGPUDescriptorSet` as long as its not a zombie.

Otherwise if the flag is not there `~IGPUDescriptorSet` can't free itself, but the pool should `reset` itself when the last created `IGPUDescriptorSet` decides to kill itself.

Finally its best if that a zombie `IGPUDescriptorSet` changes its `m_pool` back reference to `nullptr` to prevent the pool staying alive.

Reply:
I think all of this is resolved now.